pybossa/jobs.py

```diff
@@ -1297,6 +1321,46 @@ def get_management_dashboard_stats(user_email):
     send_mail(mail_dict)
 
 
+def check_and_send_task_notifications(project_id, conn=None):
+    from pybossa.core import project_repo
+
+    project = project_repo.get(project_id)
+    if not project:
+        return
+
+    reminder = project.info.get('progress_reminder', {})
+    target_remaining = reminder.get("target_remaining")
+    email_already_sent = reminder.get("sent") or False
+    if target_remaining is None:
+        return
+
+    n_remaining_tasks = n_available_tasks(project.id)
+
+    update_reminder = False
+    if n_remaining_tasks > target_remaining and email_already_sent:
+        current_app.logger.info('incomplete tasks increase over target remaining, reset reminder to active')
+        reminder['sent'] = False
+        update_reminder = True
+
+    if n_remaining_tasks <= target_remaining and not email_already_sent:
+        # incomplete tasks drop to or below, and email not sent yet, send email
+        current_app.logger.info('incomplete tasks drop, send task notification to coowners')
+        email_addr = [cached_users.get_user_email(user_id)
+                        for user_id in project.owners_ids]
+        info = dict(project_name=project.name,
+                    n_available_tasks=n_remaining_tasks)
+        notify_task_progress(info, email_addr)
+        reminder['sent'] = True
+        update_reminder = True
+
+    if update_reminder:
+        project.info['progress_reminder'] = reminder
+        if conn is not None:
+            sql = text(''' UPDATE project SET info=:info WHERE id=:id''')
+            conn.execute(sql, dict(info=json.dumps(project.info), id=project_id))
+        else:
+            project_repo.save(project)
```

Comment:
this could be audit logged

Reply:
Should we audit log when gigwork updates `sent` OR when owners update task notification (aka: progress reminder) configuration? I assume we want to update the latter?

Reply:
It could be helpful for debugging to log both scenarios.

By the way, in what case would `conn` be `None` (is that referring to a condition where the user is updating the project settings from the UI versus an automatic condition where the number of remaining tasks has changed? In both cases, it's probably a good idea to log. Thanks.

Reply:
Correct, `conn = None` happens when user update settings from UI, `conn != None` happens when listener listens to some events (e.g. update task redundancy or user submit tasks). will add logs.

Reply:
Could we add a comment in both blocks of the "if" statement to indicate this? *Just to help us remember if we come across this code in the future :)*

For example:

```python
if conn is not None:
    # Listener process is updating the task notification.
    ...
else:
    # User is updating the task notification from the project settings.
    project_repo.save(project)
```

Reply:
Sure. Btw, adding logs are not permitted here, since writing logs to db needs another db connection, concurrent db connections are prohibited. We can skip this audit log since it is not triggered by user (I checked audit logs in other project, they only track user activity)

Reply:
wii audit log when user explicitly updates task notification configuration through UI.

Reply:
sounds good